pkgs/development/python-modules/aiohttp-json-rpc/default.nix

```
@@ -0,0 +1,37 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchPypi,
+  setuptools,
+  wheel,
+  aiohttp,
+}:
+
+buildPythonPackage rec {
+  pname = "aiohttp-json-rpc";
+  version = "0.13.3";
+  pyproject = true;
+
+  src = fetchPypi {
+    inherit pname version;
+    hash = "sha256-YjehBEeMIsbvlsciegHWgyWXtBTkt5pS2FWTNWoWnpk=";
+  };
+
+  nativeBuildInputs = [
+    setuptools
+    wheel
+  ];
+
+  propagatedBuildInputs = [
+    aiohttp
+  ];
+
```

```suggestion
  nativeCheckInputs = [
  	pytestCheckHook
  ];
```
To ensure the tests are run.

---

The tests seem pretty broken on Python 3.12, using outdated attributes and parameters. Should I just close this since the repo is archived?

<details>
<summary>For reference, this is what tests are giving</summary>

```
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing pypa-build-hook
Using pypaBuildPhase
Sourcing python-runtime-deps-check-hook
Using pythonRuntimeDepsCheckHook
Sourcing pypa-install-hook
Using pypaInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing pytest-check-hook
Using pytestCheckPhase
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/jpwfs56ipfri0gq2kh0jwwcfbp98ad5n-source
source root is source
setting SOURCE_DATE_EPOCH to timestamp 315619200 of file source/tox.ini
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
no configure script, doing nothing
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
Executing pypaBuildPhase
Creating a wheel...
pypa build flags: --no-isolation --outdir dist/ --wheel
* Getting build dependencies for wheel...
running egg_info
creating aiohttp_json_rpc.egg-info
writing aiohttp_json_rpc.egg-info/PKG-INFO
writing dependency_links to aiohttp_json_rpc.egg-info/dependency_links.txt
writing entry points to aiohttp_json_rpc.egg-info/entry_points.txt
writing requirements to aiohttp_json_rpc.egg-info/requires.txt
writing top-level names to aiohttp_json_rpc.egg-info/top_level.txt
writing manifest file 'aiohttp_json_rpc.egg-info/SOURCES.txt'
reading manifest file 'aiohttp_json_rpc.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
adding license file 'LICENSE.txt'
writing manifest file 'aiohttp_json_rpc.egg-info/SOURCES.txt'
* Building wheel...
running bdist_wheel
running build
running build_py
creating build/lib/aiohttp_json_rpc
copying aiohttp_json_rpc/protocol.py -> build/lib/aiohttp_json_rpc
copying aiohttp_json_rpc/client.py -> build/lib/aiohttp_json_rpc
copying aiohttp_json_rpc/__init__.py -> build/lib/aiohttp_json_rpc
copying aiohttp_json_rpc/threading.py -> build/lib/aiohttp_json_rpc
copying aiohttp_json_rpc/communicaton.py -> build/lib/aiohttp_json_rpc
copying aiohttp_json_rpc/rpc.py -> build/lib/aiohttp_json_rpc
copying aiohttp_json_rpc/pytest.py -> build/lib/aiohttp_json_rpc
copying aiohttp_json_rpc/exceptions.py -> build/lib/aiohttp_json_rpc
copying aiohttp_json_rpc/decorators.py -> build/lib/aiohttp_json_rpc
creating build/lib/aiohttp_json_rpc/auth
copying aiohttp_json_rpc/auth/django.py -> build/lib/aiohttp_json_rpc/auth
copying aiohttp_json_rpc/auth/__init__.py -> build/lib/aiohttp_json_rpc/auth
copying aiohttp_json_rpc/auth/passwd.py -> build/lib/aiohttp_json_rpc/auth
creating build/lib/aiohttp_json_rpc/django
copying aiohttp_json_rpc/django/models.py -> build/lib/aiohttp_json_rpc/django
copying aiohttp_json_rpc/django/__init__.py -> build/lib/aiohttp_json_rpc/django
copying aiohttp_json_rpc/django/apps.py -> build/lib/aiohttp_json_rpc/django
copying aiohttp_json_rpc/django/utils.py -> build/lib/aiohttp_json_rpc/django
running egg_info
writing aiohttp_json_rpc.egg-info/PKG-INFO
writing dependency_links to aiohttp_json_rpc.egg-info/dependency_links.txt
writing entry points to aiohttp_json_rpc.egg-info/entry_points.txt
writing requirements to aiohttp_json_rpc.egg-info/requires.txt
writing top-level names to aiohttp_json_rpc.egg-info/top_level.txt
reading manifest file 'aiohttp_json_rpc.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
adding license file 'LICENSE.txt'
writing manifest file 'aiohttp_json_rpc.egg-info/SOURCES.txt'
/nix/store/6j83w4c2qdigqia0vwx14d5zzg3q471y-python3.12-setuptools-75.1.0/lib/python3.12/site-packages/setuptools/command/build_py.py:218: _Warning: Package 'aiohttp_json_rpc.django.static' is absent from the `packages` configuration.
!!

        ********************************************************************************
        ############################
        # Package would be ignored #
        ############################
        Python recognizes 'aiohttp_json_rpc.django.static' as an importable package[^1],
        but it is absent from setuptools' `packages` configuration.

        This leads to an ambiguous overall configuration. If you want to distribute this
        package, please make sure that 'aiohttp_json_rpc.django.static' is explicitly added
        to the `packages` configuration field.

        Alternatively, you can also rely on setuptools' discovery methods
        (for example by using `find_namespace_packages(...)`/`find_namespace:`
        instead of `find_packages(...)`/`find:`).

        You can read more about "package discovery" on setuptools documentation page:

        - https://setuptools.pypa.io/en/latest/userguide/package_discovery.html

        If you don't want 'aiohttp_json_rpc.django.static' to be distributed and are
        already explicitly excluding 'aiohttp_json_rpc.django.static' via
        `find_namespace_packages(...)/find_namespace` or `find_packages(...)/find`,
        you can try to use `exclude_package_data`, or `include-package-data=False` in
        combination with a more fine grained `package-data` configuration.

        You can read more about "package data files" on setuptools documentation page:

        - https://setuptools.pypa.io/en/latest/userguide/datafiles.html


        [^1]: For Python, any directory (with suitable naming) can be imported,
              even if it does not contain any `.py` files.
              On the other hand, currently there is no concept of package data
              directory, all directories are treated like packages.
        ********************************************************************************

!!
  check.warn(importable)
creating build/lib/aiohttp_json_rpc/django/static
copying aiohttp_json_rpc/django/static/rpc.js -> build/lib/aiohttp_json_rpc/django/static
installing to build/bdist.macosx-11.0-arm64/wheel
running install
running install_lib
creating build/bdist.macosx-11.0-arm64/wheel
creating build/bdist.macosx-11.0-arm64/wheel/aiohttp_json_rpc
creating build/bdist.macosx-11.0-arm64/wheel/aiohttp_json_rpc/auth
copying build/lib/aiohttp_json_rpc/auth/django.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc/auth
copying build/lib/aiohttp_json_rpc/auth/__init__.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc/auth
copying build/lib/aiohttp_json_rpc/auth/passwd.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc/auth
copying build/lib/aiohttp_json_rpc/protocol.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc
copying build/lib/aiohttp_json_rpc/client.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc
copying build/lib/aiohttp_json_rpc/__init__.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc
copying build/lib/aiohttp_json_rpc/threading.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc
copying build/lib/aiohttp_json_rpc/communicaton.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc
copying build/lib/aiohttp_json_rpc/rpc.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc
copying build/lib/aiohttp_json_rpc/pytest.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc
copying build/lib/aiohttp_json_rpc/exceptions.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc
creating build/bdist.macosx-11.0-arm64/wheel/aiohttp_json_rpc/django
copying build/lib/aiohttp_json_rpc/django/models.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc/django
copying build/lib/aiohttp_json_rpc/django/__init__.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc/django
copying build/lib/aiohttp_json_rpc/django/apps.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc/django
copying build/lib/aiohttp_json_rpc/django/utils.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc/django
creating build/bdist.macosx-11.0-arm64/wheel/aiohttp_json_rpc/django/static
copying build/lib/aiohttp_json_rpc/django/static/rpc.js -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc/django/static
copying build/lib/aiohttp_json_rpc/decorators.py -> build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc
running install_egg_info
Copying aiohttp_json_rpc.egg-info to build/bdist.macosx-11.0-arm64/wheel/./aiohttp_json_rpc-0.13.3-py3.12.egg-info
running install_scripts
creating build/bdist.macosx-11.0-arm64/wheel/aiohttp_json_rpc-0.13.3.dist-info/WHEEL
creating '/private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/dist/.tmp-3l690uw4/aiohttp_json_rpc-0.13.3-py3-none-any.whl' and adding 'build/bdist.macosx-11.0-arm64/wheel' to it
adding 'aiohttp_json_rpc/__init__.py'
adding 'aiohttp_json_rpc/client.py'
adding 'aiohttp_json_rpc/communicaton.py'
adding 'aiohttp_json_rpc/decorators.py'
adding 'aiohttp_json_rpc/exceptions.py'
adding 'aiohttp_json_rpc/protocol.py'
adding 'aiohttp_json_rpc/pytest.py'
adding 'aiohttp_json_rpc/rpc.py'
adding 'aiohttp_json_rpc/threading.py'
adding 'aiohttp_json_rpc/auth/__init__.py'
adding 'aiohttp_json_rpc/auth/django.py'
adding 'aiohttp_json_rpc/auth/passwd.py'
adding 'aiohttp_json_rpc/django/__init__.py'
adding 'aiohttp_json_rpc/django/apps.py'
adding 'aiohttp_json_rpc/django/models.py'
adding 'aiohttp_json_rpc/django/utils.py'
adding 'aiohttp_json_rpc/django/static/rpc.js'
adding 'aiohttp_json_rpc-0.13.3.dist-info/LICENSE.txt'
adding 'aiohttp_json_rpc-0.13.3.dist-info/METADATA'
adding 'aiohttp_json_rpc-0.13.3.dist-info/WHEEL'
adding 'aiohttp_json_rpc-0.13.3.dist-info/entry_points.txt'
adding 'aiohttp_json_rpc-0.13.3.dist-info/top_level.txt'
adding 'aiohttp_json_rpc-0.13.3.dist-info/RECORD'
removing build/bdist.macosx-11.0-arm64/wheel
Successfully built aiohttp_json_rpc-0.13.3-py3-none-any.whl
Finished creating a wheel...
Finished executing pypaBuildPhase
Running phase: pythonRuntimeDepsCheckHook
@nix { "action": "setPhase", "phase": "pythonRuntimeDepsCheckHook" }
Executing pythonRuntimeDepsCheck
Checking runtime dependencies for aiohttp_json_rpc-0.13.3-py3-none-any.whl
Finished executing pythonRuntimeDepsCheck
Running phase: installPhase
@nix { "action": "setPhase", "phase": "installPhase" }
Executing pypaInstallPhase
Successfully installed aiohttp_json_rpc-0.13.3-py3-none-any.whl
Finished executing pypaInstallPhase
Running phase: pythonOutputDistPhase
@nix { "action": "setPhase", "phase": "pythonOutputDistPhase" }
Executing pythonOutputDistPhase
Finished executing pythonOutputDistPhase
Running phase: fixupPhase
@nix { "action": "setPhase", "phase": "fixupPhase" }
checking for references to /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/ in /nix/store/5d29fyxhnhfgdy363ld2jlb4x73f8pi2-python3.12-aiohttp-json-rpc-0.13.3...
patching script interpreter paths in /nix/store/5d29fyxhnhfgdy363ld2jlb4x73f8pi2-python3.12-aiohttp-json-rpc-0.13.3
stripping (with command strip and flags -S) in  /nix/store/5d29fyxhnhfgdy363ld2jlb4x73f8pi2-python3.12-aiohttp-json-rpc-0.13.3/lib
checking for references to /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/ in /nix/store/fjr6w33faxj303zivq1pn9i52pw8djlf-python3.12-aiohttp-json-rpc-0.13.3-dist...
patching script interpreter paths in /nix/store/fjr6w33faxj303zivq1pn9i52pw8djlf-python3.12-aiohttp-json-rpc-0.13.3-dist
Executing pythonRemoveTestsDir
Finished executing pythonRemoveTestsDir
Running phase: installCheckPhase
@nix { "action": "setPhase", "phase": "installCheckPhase" }
no Makefile or custom installCheckPhase, doing nothing
Running phase: pythonCatchConflictsPhase
@nix { "action": "setPhase", "phase": "pythonCatchConflictsPhase" }
Running phase: pythonRemoveBinBytecodePhase
@nix { "action": "setPhase", "phase": "pythonRemoveBinBytecodePhase" }
Running phase: pythonImportsCheckPhase
@nix { "action": "setPhase", "phase": "pythonImportsCheckPhase" }
Executing pythonImportsCheckPhase
Check whether the following modules can be imported: aiohttp_json_rpc
Running phase: pytestCheckPhase
@nix { "action": "setPhase", "phase": "pytestCheckPhase" }
Executing pytestCheckPhase
============================= test session starts ==============================
platform darwin -- Python 3.12.7, pytest-8.3.3, pluggy-1.5.0 -- /nix/store/zyak8iqzh1ww83qa4sqwwz3qax0lrky7-python3-3.12.7/bin/python3.12
cachedir: .pytest_cache
rootdir: /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source
configfile: pytest.ini
plugins: aiohttp-json-rpc-0.13.3, asyncio-0.23.8
asyncio: mode=Mode.STRICT
collecting ... 
collected 41 items                                                             

tests/test_client.py::test_client_connect_disconnect PASSED              [  2%]
tests/test_client.py::test_client_autoconnect PASSED                     [  4%]
tests/test_client.py::test_client_connection_failure PASSED              [  7%]
tests/test_decorators.py::test_raw_responses PASSED                      [  9%]
tests/test_django_auth_backend.py::test_login ERROR                      [ 12%]
tests/test_django_auth_backend.py::test_generic_orm_methods ERROR        [ 14%]
tests/test_django_auth_backend.py::test_generic_orm_view ERROR           [ 17%]
tests/test_django_auth_backend.py::test_generic_orm_delete ERROR         [ 19%]
tests/test_django_auth_backend.py::test_generic_orm_add ERROR            [ 21%]
tests/test_django_auth_backend.py::test_generic_orm_change ERROR         [ 24%]
tests/test_django_transactions.py::test_cuncurrent_transactions FAILED   [ 26%]
tests/test_errors.py::test_RpcMethodNotFoundError PASSED                 [ 29%]
tests/test_errors.py::test_RpcInvalidRequestError PASSED                 [ 31%]
tests/test_errors.py::test_RpcInvalidParamsError PASSED                  [ 34%]
tests/test_errors.py::test_RpcGenericServerDefinedError_raw PASSED       [ 36%]
tests/test_errors.py::test_RpcGenericServerDefinedError PASSED           [ 39%]
tests/test_forms.py::test_confirmation PASSED                            [ 41%]
tests/test_message_decoding.py::test_valid_message PASSED                [ 43%]
tests/test_message_decoding.py::test_valid_notification PASSED           [ 46%]
tests/test_message_decoding.py::test_valid_error PASSED                  [ 48%]
tests/test_message_decoding.py::test_invalid_error PASSED                [ 51%]
tests/test_method_api.py::test_async_notification_api PASSED             [ 53%]
tests/test_method_api.py::test_sync_notification_api PASSED              [ 56%]
tests/test_method_api.py::test_async_method_api PASSED                   [ 58%]
tests/test_method_api.py::test_sync_method_api PASSED                    [ 60%]
tests/test_method_args.py::test_args PASSED                              [ 63%]
tests/test_method_args.py::test_to_few_arguments PASSED                  [ 65%]
tests/test_method_args.py::test_type_validators PASSED                   [ 68%]
tests/test_method_args.py::test_function_validators PASSED               [ 70%]
tests/test_method_args.py::test_introspection_of_unsupported_function_types PASSED [ 73%]
tests/test_methods.py::test_get_methods PASSED                           [ 75%]
tests/test_methods.py::test_add_methods PASSED                           [ 78%]
tests/test_methods.py::test_add_methods_diff_prefixes PASSED             [ 80%]
tests/test_methods.py::test_call_method PASSED                           [ 82%]
tests/test_methods.py::test_method_names PASSED                          [ 85%]
tests/test_publish_subscribe.py::test_add_topics PASSED                  [ 87%]
tests/test_publish_subscribe.py::test_subscribe PASSED                   [ 90%]
tests/test_publish_subscribe.py::test_unsubscribe PASSED                 [ 92%]
tests/test_publish_subscribe.py::test_notify PASSED                      [ 95%]
tests/test_publish_subscribe.py::test_state PASSED                       [ 97%]
tests/test_publish_subscribe.py::test_notify_on_request FAILED           [100%]

==================================== ERRORS ====================================
_________________________ ERROR at setup of test_login _________________________
file /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/tests/test_django_auth_backend.py, line 14
  @pytest.mark.asyncio
  async def test_login(django_rpc_context, django_staff_user):
      from aiohttp_json_rpc.auth import login_required

      # setup rpc
      @login_required
      async def restricted_method(request):
          return True

      django_rpc_context.rpc.add_methods(('', restricted_method))

      # setup client
      client = await django_rpc_context.make_client()

      # run test
      # # without login
      methods = await client.call('get_methods')

      assert 'login' in methods
      assert 'restricted_method' not in methods

      # # after login login
      assert await client.call('login', {
          'username': 'admin',
          'password': 'admin',
      })

      methods = await client.call('get_methods')

      assert 'login' not in methods
      assert 'restricted_method' in methods
file /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/pytest.py, line 101
  @pytest.yield_fixture
  def django_rpc_context(db, event_loop, unused_tcp_port):
E       fixture 'db' not found
>       available fixtures: _session_event_loop, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, django_rpc_context, django_staff_user, doctest_namespace, event_loop, event_loop_policy, items, monkeypatch, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, rpc_context, tests/test_django_auth_backend.py::<event_loop>, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory, unused_tcp_port, unused_tcp_port_factory, unused_udp_port, unused_udp_port_factory
>       use 'pytest --fixtures [testpath]' for help on them.

/private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/pytest.py:101
__________________ ERROR at setup of test_generic_orm_methods __________________
file /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/tests/test_django_auth_backend.py, line 47
  @pytest.mark.asyncio
  async def test_generic_orm_methods(django_rpc_context, django_staff_user):
      client = await django_rpc_context.make_client()

      assert await client.call('login', {
          'username': 'admin',
          'password': 'admin',
      })

      methods = await client.call('get_methods')

      assert list(filter(lambda m: m.startswith('db__'), methods))
file /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/pytest.py, line 101
  @pytest.yield_fixture
  def django_rpc_context(db, event_loop, unused_tcp_port):
E       fixture 'db' not found
>       available fixtures: _session_event_loop, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, django_rpc_context, django_staff_user, doctest_namespace, event_loop, event_loop_policy, items, monkeypatch, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, rpc_context, tests/test_django_auth_backend.py::<event_loop>, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory, unused_tcp_port, unused_tcp_port_factory, unused_udp_port, unused_udp_port_factory
>       use 'pytest --fixtures [testpath]' for help on them.

/private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/pytest.py:101
___________________ ERROR at setup of test_generic_orm_view ____________________
file /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/tests/test_django_auth_backend.py, line 61
  @pytest.mark.asyncio
  async def test_generic_orm_view(django_rpc_context, django_staff_user, items):
      client = await django_rpc_context.make_client()

      assert await client.call('login', {
          'username': 'admin',
          'password': 'admin',
      })

      assert len(await client.call('db__django_project.view_item')) == 10

      items = await client.call('db__django_project.view_item', {
          'number__gt': 0,
          'number__lt': 9,
      })

      assert len(items) == 8
file /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/pytest.py, line 101
  @pytest.yield_fixture
  def django_rpc_context(db, event_loop, unused_tcp_port):
E       fixture 'db' not found
>       available fixtures: _session_event_loop, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, django_rpc_context, django_staff_user, doctest_namespace, event_loop, event_loop_policy, items, monkeypatch, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, rpc_context, tests/test_django_auth_backend.py::<event_loop>, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory, unused_tcp_port, unused_tcp_port_factory, unused_udp_port, unused_udp_port_factory
>       use 'pytest --fixtures [testpath]' for help on them.

/private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/pytest.py:101
__________________ ERROR at setup of test_generic_orm_delete ___________________
file /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/tests/test_django_auth_backend.py, line 80
  @pytest.mark.asyncio
  async def test_generic_orm_delete(django_rpc_context, django_staff_user,
                                    items):

      client = await django_rpc_context.make_client()

      assert await client.call('login', {
          'username': 'admin',
          'password': 'admin',
      })

      assert len(await client.call('db__django_project.view_item')) == 10

      await client.call('db__django_project.delete_item', {
          'number__gt': 0,
          'number__lt': 9,
      })

      assert len(await client.call('db__django_project.view_item')) == 2
file /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/pytest.py, line 101
  @pytest.yield_fixture
  def django_rpc_context(db, event_loop, unused_tcp_port):
E       fixture 'db' not found
>       available fixtures: _session_event_loop, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, django_rpc_context, django_staff_user, doctest_namespace, event_loop, event_loop_policy, items, monkeypatch, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, rpc_context, tests/test_django_auth_backend.py::<event_loop>, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory, unused_tcp_port, unused_tcp_port_factory, unused_udp_port, unused_udp_port_factory
>       use 'pytest --fixtures [testpath]' for help on them.

/private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/pytest.py:101
____________________ ERROR at setup of test_generic_orm_add ____________________
file /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/tests/test_django_auth_backend.py, line 101
  @pytest.mark.asyncio
  async def test_generic_orm_add(django_rpc_context, django_staff_user, items):
      client = await django_rpc_context.make_client()

      assert await client.call('login', {
          'username': 'admin',
          'password': 'admin',
      })

      assert len(await client.call('db__django_project.view_item')) == 10

      new_item = await client.call('db__django_project.add_item', {
          'client_id': 100,
          'number': 100,
      })

      assert len(await client.call('db__django_project.view_item')) == 11

      assert await client.call('db__django_project.view_item', {
          'pk': new_item['pk'],
      })
file /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/pytest.py, line 101
  @pytest.yield_fixture
  def django_rpc_context(db, event_loop, unused_tcp_port):
E       fixture 'db' not found
>       available fixtures: _session_event_loop, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, django_rpc_context, django_staff_user, doctest_namespace, event_loop, event_loop_policy, items, monkeypatch, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, rpc_context, tests/test_django_auth_backend.py::<event_loop>, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory, unused_tcp_port, unused_tcp_port_factory, unused_udp_port, unused_udp_port_factory
>       use 'pytest --fixtures [testpath]' for help on them.

/private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/pytest.py:101
__________________ ERROR at setup of test_generic_orm_change ___________________
file /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/tests/test_django_auth_backend.py, line 124
  @pytest.mark.asyncio
  async def test_generic_orm_change(django_rpc_context, django_staff_user,
                                    items):

      client = await django_rpc_context.make_client()

      assert await client.call('login', {
          'username': 'admin',
          'password': 'admin',
      })

      item = (await client.call('db__django_project.view_item'))[0]

      assert await client.call('db__django_project.change_item', {
          'pk': item['pk'],
          'number': item['number'] + 1,
      })

      change_item = (await client.call('db__django_project.view_item', {
          'pk': item['pk'],
      }))[0]

      assert change_item['number'] == item['number'] + 1
file /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/pytest.py, line 101
  @pytest.yield_fixture
  def django_rpc_context(db, event_loop, unused_tcp_port):
E       fixture 'db' not found
>       available fixtures: _session_event_loop, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, django_rpc_context, django_staff_user, doctest_namespace, event_loop, event_loop_policy, items, monkeypatch, pytestconfig, record_property, record_testsuite_property, record_xml_attribute, recwarn, rpc_context, tests/test_django_auth_backend.py::<event_loop>, tmp_path, tmp_path_factory, tmpdir, tmpdir_factory, unused_tcp_port, unused_tcp_port_factory, unused_udp_port, unused_udp_port_factory
>       use 'pytest --fixtures [testpath]' for help on them.

/private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/pytest.py:101
=================================== FAILURES ===================================
_________________________ test_cuncurrent_transactions _________________________

event_loop = <_UnixSelectorEventLoop running=False closed=False debug=False>
unused_tcp_port = 63897

    @pytest.mark.django_db(transaction=True)
    @pytest.mark.asyncio
    async def test_cuncurrent_transactions(event_loop, unused_tcp_port):
        """
        This test tests aiohttp_json_rpc.django.patch_db_connections.
    
        The test sets up a JsonRpc and starts two concurrent clients.
        Both clients try to open a Django transaction.
        The test is successful if only the first client is able to commit its
        database changes.
    
        Note:
            The transactions are slowed down artificial and Client #2 sleeps 200ms
            on start using asyncio.sleep to make the test log more readable and
            ensure the transactions are concurrent.
    
        Note:
            The test starts an watchdog to make sure the test does not hang if the
            rpc communication is broken or hanging.
        """
    
>       from django_project.models import Item
E       ModuleNotFoundError: No module named 'django_project.models'

tests/test_django_transactions.py:135: ModuleNotFoundError
------------------------------ Captured log setup ------------------------------
DEBUG    asyncio                        [17:06:02.891] Using selector: KqueueSelector
---------------------------- Captured log teardown -----------------------------
DEBUG    asyncio                        [17:06:02.902] Using selector: KqueueSelector
____________________________ test_notify_on_request ____________________________

rpc_context = <aiohttp_json_rpc.pytest.RpcContext object at 0x103653860>
event_loop = <_UnixSelectorEventLoop running=False closed=False debug=False>

    @pytest.mark.asyncio
    async def test_notify_on_request(rpc_context, event_loop):
        import asyncio
    
        # setup rpc
        rpc_context.rpc.add_topics('topic')
    
        async def method(request):
            await request.send_notification('topic', 'foo')
    
            return True
    
        rpc_context.rpc.add_methods(('', method))
    
        # setup client
        message = asyncio.Future()
    
        async def handler(data):
            if data['method'] == 'topic':
                message.set_result(data)
    
        client = await rpc_context.make_client()
        await client.subscribe('topic', handler)
    
        # run test
>       await asyncio.gather(
            client.call('method'),
            asyncio.wait_for(message, 1),
            loop=event_loop,
        )
E       TypeError: gather() got an unexpected keyword argument 'loop'

tests/test_publish_subscribe.py:124: TypeError
------------------------------ Captured log setup ------------------------------
DEBUG    asyncio                        [17:06:03.068] Using selector: KqueueSelector
------------------------------ Captured log call -------------------------------
DEBUG    aiohttp-json-rpc.client        [17:06:03.069] #31: ws connect...
ERROR    asyncio                        [17:06:03.069] Task exception was never retrieved
future: <Task finished name='Task-516' coro=<JsonRpcClient._handle_msgs() done, defined at /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/client.py:64> exception=AttributeError("'JsonRpcClient' object has no attribute '_ws'")>
Traceback (most recent call last):
  File "/private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/client.py", line 67, in _handle_msgs
    while not self._ws.closed:
              ^^^^^^^^
AttributeError: 'JsonRpcClient' object has no attribute '_ws'
DEBUG    aiohttp-json-rpc.server        [17:06:03.073] waiting for messages
DEBUG    aiohttp-json-rpc.client        [17:06:03.073] #31: ws connected
DEBUG    aiohttp-json-rpc.client        [17:06:03.073] #31: worker start...
DEBUG    aiohttp-json-rpc.client        [17:06:03.073] #31: waiting for messages
DEBUG    aiohttp-json-rpc.client        [17:06:03.073] #31: > {"jsonrpc": "2.0", "method": "subscribe", "id": 0, "params": "topic"}
DEBUG    aiohttp-json-rpc.server        [17:06:03.073] raw msg received: {"jsonrpc": "2.0", "method": "subscribe", "id": 0, "params": "topic"}
DEBUG    aiohttp-json-rpc.server        [17:06:03.073] waiting for messages
DEBUG    aiohttp-json-rpc.server        [17:06:03.073] message decoded: JsonRpcMsg(type=10, data={'jsonrpc': '2.0', 'method': 'subscribe', 'id': 0, 'params': 'topic'})
DEBUG    aiohttp-json-rpc.server        [17:06:03.073] msg gets handled as request
DEBUG    aiohttp-json-rpc.client        [17:06:03.073] #31: < {"jsonrpc": "2.0", "id": 0, "result": ["topic"]}
DEBUG    aiohttp-json-rpc.client        [17:06:03.073] #31: waiting for messages
---------------------------- Captured log teardown -----------------------------
DEBUG    aiohttp-json-rpc.client        [17:06:03.075] #31: < None
DEBUG    aiohttp-json-rpc.client        [17:06:03.075] #31: waiting for messages
DEBUG    aiohttp-json-rpc.client        [17:06:03.075] #31: < None
INFO     aiohttp.access                 [17:06:03.075] ::1 [18/Dec/2024:17:06:03 +0000] "GET /rpc HTTP/1.1" 101 0 "-" "Python/3.12 aiohttp/3.10.10"
DEBUG    asyncio                        [17:06:03.075] Using selector: KqueueSelector
=============================== warnings summary ===============================
tests/test_django_auth_backend.py:3
  /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/tests/test_django_auth_backend.py:3: PytestUnknownMarkWarning: Unknown pytest.mark.django - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    pytestmark = pytest.mark.django(reason='Depends on Django')

tests/test_django_transactions.py:12
  /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/tests/test_django_transactions.py:12: PytestUnknownMarkWarning: Unknown pytest.mark.django - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    pytestmark = pytest.mark.django(reason='Depends on Django')

tests/test_django_transactions.py:114
  /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/tests/test_django_transactions.py:114: PytestUnknownMarkWarning: Unknown pytest.mark.django_db - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.django_db(transaction=True)

tests/test_django_transactions.py:114
  tests/test_django_transactions.py:114: PytestDeprecationWarning: test_cuncurrent_transactions is asynchronous and explicitly requests the "event_loop" fixture. Asynchronous fixtures and test functions should use "asyncio.get_running_loop()" instead.
    @pytest.mark.django_db(transaction=True)

tests/test_publish_subscribe.py:99
  tests/test_publish_subscribe.py:99: PytestDeprecationWarning: test_notify_on_request is asynchronous and explicitly requests the "event_loop" fixture. Asynchronous fixtures and test functions should use "asyncio.get_running_loop()" instead.
    @pytest.mark.asyncio

tests/test_client.py: 4 warnings
tests/test_decorators.py: 1 warning
tests/test_errors.py: 5 warnings
tests/test_forms.py: 1 warning
tests/test_method_api.py: 4 warnings
tests/test_method_args.py: 4 warnings
tests/test_methods.py: 7 warnings
tests/test_publish_subscribe.py: 7 warnings
  /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/client.py:125: DeprecationWarning: ssl=None is deprecated, please use ssl=True
    self._ws = await self._session.ws_connect(url, ssl=ssl)

tests/test_publish_subscribe.py::test_notify_on_request
  /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/tests/test_publish_subscribe.py:124: RuntimeWarning: coroutine 'wait_for' was never awaited
    await asyncio.gather(
  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.

tests/test_publish_subscribe.py::test_notify_on_request
  /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/tests/test_publish_subscribe.py:124: RuntimeWarning: coroutine 'JsonRpcClient.call' was never awaited
    await asyncio.gather(
  Enable tracemalloc to get traceback where the object was allocated.
  See https://docs.pytest.org/en/stable/how-to/capture-warnings.html#resource-warnings for more info.

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED tests/test_django_transactions.py::test_cuncurrent_transactions - ModuleNotFoundError: No module named 'django_project.models'
FAILED tests/test_publish_subscribe.py::test_notify_on_request - TypeError: gather() got an unexpected keyword argument 'loop'
ERROR tests/test_django_auth_backend.py::test_login
ERROR tests/test_django_auth_backend.py::test_generic_orm_methods
ERROR tests/test_django_auth_backend.py::test_generic_orm_view
ERROR tests/test_django_auth_backend.py::test_generic_orm_delete
ERROR tests/test_django_auth_backend.py::test_generic_orm_add
ERROR tests/test_django_auth_backend.py::test_generic_orm_change
============= 2 failed, 33 passed, 40 warnings, 6 errors in 0.28s ==============
Task exception was never retrieved
future: <Task finished name='Task-107' coro=<JsonRpcClient._handle_msgs() done, defined at /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/client.py:64> exception=AttributeError("'JsonRpcClient' object has no attribute '_ws'")>
Traceback (most recent call last):
  File "/private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/client.py", line 67, in _handle_msgs
    while not self._ws.closed:
              ^^^^^^^^
AttributeError: 'JsonRpcClient' object has no attribute '_ws'
Task exception was never retrieved
future: <Task finished name='Task-419' coro=<JsonRpcClient._handle_msgs() done, defined at /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/client.py:64> exception=AttributeError("'JsonRpcClient' object has no attribute '_ws'")>
Traceback (most recent call last):
  File "/private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/client.py", line 67, in _handle_msgs
    while not self._ws.closed:
              ^^^^^^^^
AttributeError: 'JsonRpcClient' object has no attribute '_ws'
Task exception was never retrieved
future: <Task finished name='Task-437' coro=<JsonRpcClient._handle_msgs() done, defined at /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/client.py:64> exception=AttributeError("'JsonRpcClient' object has no attribute '_ws'")>
Traceback (most recent call last):
  File "/private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/client.py", line 67, in _handle_msgs
    while not self._ws.closed:
              ^^^^^^^^
AttributeError: 'JsonRpcClient' object has no attribute '_ws'
Task exception was never retrieved
future: <Task finished name='Task-446' coro=<JsonRpcClient._handle_msgs() done, defined at /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/client.py:64> exception=AttributeError("'JsonRpcClient' object has no attribute '_ws'")>
Traceback (most recent call last):
  File "/private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/client.py", line 67, in _handle_msgs
    while not self._ws.closed:
              ^^^^^^^^
AttributeError: 'JsonRpcClient' object has no attribute '_ws'
Task exception was never retrieved
future: <Task finished name='Task-481' coro=<JsonRpcClient._handle_msgs() done, defined at /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/client.py:64> exception=AttributeError("'JsonRpcClient' object has no attribute '_ws'")>
Traceback (most recent call last):
  File "/private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/client.py", line 67, in _handle_msgs
    while not self._ws.closed:
              ^^^^^^^^
AttributeError: 'JsonRpcClient' object has no attribute '_ws'
Task exception was never retrieved
future: <Task finished name='Task-500' coro=<JsonRpcClient._handle_msgs() done, defined at /private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/client.py:64> exception=AttributeError("'JsonRpcClient' object has no attribute '_ws'")>
Traceback (most recent call last):
  File "/private/tmp/nix-build-python3.12-aiohttp-json-rpc-0.13.3.drv-0/source/aiohttp_json_rpc/client.py", line 67, in _handle_msgs
    while not self._ws.closed:
              ^^^^^^^^
AttributeError: 'JsonRpcClient' object has no attribute '_ws'
```

---

In that case it seems unlikely the package will work as intended. You can close the PR.